### DISCUSSION TITLE
How to hide direct observable property access?

### DISCUSSION THREAD

**Discussion by erkinsergey** [mobxjs/mobx#3506] [Q&A] [answered]
I want to check the observable property value being set and, for example, log changes. For this I use a getter and a setter.
```
const store = makeAutoObservable({
  __some: 'value',
 set some(v) {
   console.log(`New value of some is ${v}`);
   this.__some = v;
 },
 get some() {
   return this._some;
 }
});

autorun(() => console.log(`Value of some is ${store.some}`));
```
It works. But the property ```store.__some``` is also read/write (with warnings). 
How can I hide direct external access to an observable property while still maintaining control over it?

**Comment by urugator** [comment]
Do you mean how to make `__some` private (accessible only from getter/setter)? If yes how does it relate to `__some` being observable or not and therefore to mobx?

**Reply by erkinsergey** [reply]
> Do you mean how to make `__some` private (accessible only from getter/setter)? If yes how does it relate to `__some` being observable or not and therefore to mobx?

Yes, I would like to have an observable property only accessible via getters and setters.
I can work with the property _some_
```
  console.log(store.some);
  store.some = 'someNewValue';

  // console output via setter is: 
  // New value of some is someNewValue
```
But the property ___some_ is also accessible from the outside
```
  console.log(store.__some);
  store.__some = 'someNewValue';
  
  // no console output, setter didn't work
```
In both cases, these properties are observables. But in the second case, I lose control of the new value.
I'd like to remove access to ___some_ from outside, so that _some_ remains observable?

**Comment by erkinsergey** [comment, marked as answer]
After some experimentation, I found a working way to do this.
```
  /**
   * Only module private interface for observable entity
   */
  const someObservable = makeAutoObservable({
    __some: 'initialValue',
    get some() {
        return this.__some;
    },
    set some(v) {
        this.__some = v;
    },
  });

  /**
   * Public interface for all consumers
   */
  const store1 = Object.freeze({
    // The function-method only, getter not work here
    getSome() {
        return someObservable.some;
    },
    // The function-method only, setter not work here
    setSome(v) {
        // cheking conditions
        if ('number' !== typeof v) {
            return;
        }

        someObservable.some = v;
    }
  });

  /**
   * Final result with all such public interfaces
   */
  const store = Object.freeze({
    ...store1,
    // ...store2,
    // ...
  });

  autorun(() => console.log(`current value of some is ${store.getSome()}`));

  store.setSome(77);
  store.setSome('4444');

  console.log(JSON.stringify(store));
```
Output
```
current value of some is initialValue
current value of some is 77
{}
```
